lsp/Main.hs

```diff
@@ -39,29 +42,58 @@ handlers = mconcat
       pure ()
   , requestHandler STextDocumentHover $ \req responder -> do
       let RequestMessage _ _ _ (HoverParams doc pos _workDone) = req
-          Position _l _c' = pos
-          rsp = Hover ms (Just range)
-          ms = HoverContents $ markedUpContent "lsp-demo-simple-server" "Hello world"
-          range = Range pos pos
-      rsp <- lookupToken pos doc
-      -- responder (Right $ Just rsp)
+      let (TextDocumentIdentifier uri) = doc
+      let exceptHover = lookupTokenBare' pos uri
```

Comment:
This doesn't actually run the action. Instead you are running the parser twice: Once for `runDiagnosticsOnHover` and again in `runHoverResponse`. I assume this is not the intention.

Reply:
Is it because it's lazy?

Reply:
No, it's not laziness in this case. It is because there is a difference between working with a value of type `IO a` and actually executing the IO action.
You can kind of see Haskell as a pure language that allows you to build up trees of IO actions, which are then executed by a separate engine.

All that `>>=` does is chain together more such actions in the tree.

When you do `let someThing = someIoAction` it doesn't do anything more than give a new name to the action. In order to have it be executed, it needs to be inserted in the chain with `>>=` or `_ <- someThing` and each time you do that, it will run the action again.

This allows us to use functions like
```
iterateM 4 (putStrLn "Hello")
```
since we can manipulate IO actions just like normal Haskell values.

Sorry if I'm making this more confusing than helping.